File: cpptet_v0.5/CTetris.cpp
#include "CTetris.h"
#include <cmath>

Matrix **CTetris::setOfCBlockObjects;

void CTetris::init(int *setOfBlockArrays[], int nTypes, int nDegrees) {
    Tetris::init(setOfBlockArrays, nTypes, nDegrees);
    setOfCBlockObjects = new Matrix*[nBlockTypes];
    for(int y=0; y<nBlockTypes; y++)
        setOfCBlockObjects[y] = new Matrix[nBlockDegrees];
    
    int size[nBlockTypes];
    for(int i=0; i<nBlockTypes; i++)
        size[i] = 0;

    for(int i=0; i<nBlockTypes; i++) {
        for(int j=0; setOfBlockArrays[i * nBlockDegrees][j]!=-1; j++)
            size[i]++;
        size[i] = (int)sqrt(size[i]);
    }

    for(int i=0; i<nBlockTypes; i++){
        for(int j=0; j<nBlockDegrees; j++){
            Matrix obj = Matrix(setOfBlockArrays[i * nBlockDegrees + j], size[i], size[i]);
            obj.mulc(i+1);
            setOfCBlockObjects[i][j] = obj;
        }
    }
}

CTetris::CTetris(int dy, int dx) : Tetris(dy, dx){
    currCBlk = new Matrix();
    iCScreen = new Matrix(createArrayScreen(), arrayScreenDy, arrayScreenDx);
    oCScreen = new Matrix(iCScreen);
}

CTetris::~CTetris() {
    for(int i=0; i<nBlockTypes; i++)
        delete [] setOfCBlockObjects[i];
    delete [] setOfCBlockObjects;
    delete currCBlk;
    delete iCScreen;
    delete oCScreen;
}

TetrisState CTetris::accept(char key) {
    if(key>='0' && key<=('0' + nBlockTypes - 1)){
        if(justStarted == false)
            deleteFullLines();
        *iCScreen = Matrix(oCScreen);
    }

    state = Tetris::accept(key);

    *currCBlk = setOfCBlockObjects[idxBlockType][idxBlockDegree];
    Matrix* tempCBlk = iCScreen->clip(top, left, top+currCBlk->get_dy(), left+currCBlk->get_dx());
    tempCBlk = tempCBlk->add(currCBlk);

    *oCScreen = Matrix(iCScreen);
    oCScreen->paste(tempCBlk, top, left);
    return state;
}

void CTetris::deleteFullLines() {
    int **array = oScreen->get_array();
    int **cArray = oCScreen->get_array();
        
    for(int y=oScreen->get_dy()-iScreenDw-1; y>=0; y--) {
        bool isFull = true;

        for(int x=iScreenDw; x<oScreen->get_dx()-iScreenDw; x++) {
            if(array[y][x] == 0){
                isFull = false;
                break;
            }
        }
        if(isFull) {
            for(int line=y; line>0; line--) {
                for(int x=iScreenDw; x<oScreen->get_dx()-iScreenDw; x++) {
                    array[line][x] = array[line-1][x];
                    cArray[line][x] = cArray[line-1][x];
                }
            }
            for(int x=iScreenDw; x<oScreen->get_dx()-iScreenDw; x++) {
                array[0][x] = 0;
                cArray[0][x] = 0;
            }
            
            int *array2 = new int[oScreen->get_dy() * oScreen->get_dx() + 1];
            int *cArray2 = new int[oScreen->get_dy() * oScreen->get_dx() + 1];
            for(int i=0; i<oScreen->get_dy(); i++) {
                for(int j=0; j<oScreen->get_dx(); j++) {
                    array2[i * oScreen->get_dx() + j] = array[i][j];
                    cArray2[i * oScreen->get_dx() + j] = cArray[i][j];
                }
            }
            *oScreen = Matrix(array2, oScreen->get_dy(), oScreen->get_dx());
            *oCScreen = Matrix(cArray2, oScreen->get_dy(), oScreen->get_dx());
            return deleteFullLines();
        }
    }

    return;
}

File: pytet_v0.5/ctetris.py
from tetris import *

class CTetris((Tetris)):
    setOfCBlockObjects = 0
    
    @classmethod
    def init(cls, setOfBlockArrays):
        Tetris.init(setOfBlockArrays)
        CTetris.setOfCBlockObjects = [[0]* Tetris.nBlockDegrees for _ in range(Tetris.nBlockTypes)]

        for i in range(Tetris.nBlockTypes):
            for j in range(Tetris.nBlockDegrees):
                obj = Matrix(setOfBlockArrays[i][j])
                obj.mulc(i+1)
                CTetris.setOfCBlockObjects[i][j] = obj
        return

    def __init__(self, cy, cx):
        Tetris.__init__(self, cy, cx)
        arrayScreen = self.createArrayScreen()
        self.iCScreen = Matrix(arrayScreen)
        self.oCScreen = Matrix(self.iCScreen)
        return

    def accept(self, key):
        if key >= '0' and key <= '6':
            if self.justStarted == False:
                self.deleteFullLines()
            self.iCScreen = Matrix(self.oCScreen)

        state = Tetris.accept(self, key)

        currCBlk = CTetris.setOfCBlockObjects[self.idxBlockType][self.idxBlockDegree]
        tempBlk = self.iCScreen.clip(self.top, self.left,
                                            self.top + currCBlk.get_dy(),
                                            self.left + currCBlk.get_dx())
        tempBlk = tempBlk + currCBlk

        self.oCScreen = Matrix(self.iCScreen)
        self.oCScreen.paste(tempBlk, self.top, self.left)
        return state

    def deleteFullLines(self):
        array = self.oScreen.get_array()
        cArray = self.oCScreen.get_array()
        
        for y in range(self.oScreen.get_dy()-CTetris.iScreenDw-1, 0, -1):
            for x in range(CTetris.iScreenDw, self.oScreen.get_dx()-CTetris.iScreenDw):
                if array[y][x] == 0:
                    break
            else:
                for line in range(y, 0, -1):
                    array[line] = array[line-1][ : ]
                    cArray[line] = cArray[line-1][ : ]
                for x in range(CTetris.iScreenDw, self.oScreen.get_dx()-CTetris.iScreenDw):
                    array[0][x] = 0
                    cArray[0][x] = 0
                self.oScreen = Matrix(array)
                self.oCScreen = Matrix(cArray)
                return self.deleteFullLines()

        return


### end of class CTetris():
